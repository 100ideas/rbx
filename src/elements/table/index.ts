import "./table.module.sass";

export { Table } from "./table";
