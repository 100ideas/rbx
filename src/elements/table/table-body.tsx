import React from "react";

import { forwardRefAs, Generic } from "../../base";
import { HelpersProps } from "../../base/helpers";

export type TableBodyOwnProps = HelpersProps;
export type TableBodyForwardsProps = { className: string };

export const TableBody = forwardRefAs<
  "tbody",
  TableBodyOwnProps,
  TableBodyForwardsProps
>((props, ref) => <Generic ref={ref} {...props} />, { as: "tbody" });

TableBody.displayName = "Table.Body";
