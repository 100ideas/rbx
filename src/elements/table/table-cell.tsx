import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";

export type TableCellProps = HelpersProps;

export const TableCell = forwardRefAs<TableCellProps, "td">(
  (props, ref) => <Generic ref={ref} {...props} />,
  { as: "td" },
);

TableCell.displayName = "Table.Cell";
