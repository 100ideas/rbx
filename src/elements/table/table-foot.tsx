import React from "react";

import { forwardRefAs, Generic } from "../../base";
import { HelpersProps } from "../../base/helpers";

export type TableFootOwnProps = HelpersProps;
export type TableFootForwardsProps = { className: string };

export const TableFoot = forwardRefAs<
  "tfoot",
  TableFootOwnProps,
  TableFootForwardsProps
>((props, ref) => <Generic ref={ref} {...props} />, { as: "tfoot" });

TableFoot.displayName = "Table.Foot";
