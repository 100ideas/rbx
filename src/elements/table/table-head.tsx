import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";

export type TableHeadProps = HelpersProps;

export const TableHead = forwardRefAs<TableHeadProps>(
  (props, ref) => <Generic ref={ref} {...props} />,
  { as: "thead" },
);

TableHead.displayName = "Table.Head";
