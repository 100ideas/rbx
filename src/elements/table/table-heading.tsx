import React from "react";

import { forwardRefAs, Generic } from "../../base";
import { HelpersProps } from "../../base/helpers";

export type TableHeadingOwnProps = HelpersProps;
export type TableHeadingForwardsProps = { className: string };

export const TableHeading = forwardRefAs<
  "th",
  TableHeadingOwnProps,
  TableHeadingForwardsProps
>((props, ref) => <Generic ref={ref} {...props} />, { as: "th" });

TableHeading.displayName = "Table.Heading";
