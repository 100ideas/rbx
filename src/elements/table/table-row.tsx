import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";

export type TableRowHelperProps = Partial<{
  selected: boolean;
}>;

export type TableRowProps = HelpersProps & TableRowHelperProps;

const propTypes = {
  selected: PropTypes.bool,
};

export const TableRow = Object.assign(
  forwardRefAs<TableRowProps, "tr">(
    ({ className, selected, ...rest }, ref) => (
      <Generic
        className={classNames({ "is-selected": selected }, className)}
        ref={ref}
        {...rest}
      />
    ),
    { as: "tr" },
  ),
  { propTypes },
);
