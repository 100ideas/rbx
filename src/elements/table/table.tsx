import { cx } from "emotion";
import React from "react";

import { Generic } from "@/generic";
import { ModifierProps } from "@/modifiers";

export type TableModifierProps = Partial<{
  bordered: boolean;
  fullwidth: boolean;
  hoverable: boolean;
  narrow: boolean;
  striped: boolean;
}>;

export type TableProps = Prefer<
  ModifierProps & TableModifierProps,
  React.HTMLAttributes<HTMLTableElement>
>;

export const Table = React.forwardRef<HTMLTableElement, TableProps>(
  (props, ref) => {
    const { bordered, fullwidth, hoverable, narrow, striped, ...rest } = props;
    rest.className = cx("table", rest.className, {
      "is-bordered": bordered,
      "is-fullwidth": fullwidth,
      "is-hoverable": hoverable,
      "is-narrow": narrow,
      "is-striped": striped,
    });
    return <Generic<"table"> as="table" ref={ref} {...rest} />;
  },
);

Table.defaultProps = {
  bordered: false,
  fullwidth: false,
  hoverable: false,
  narrow: false,
  striped: false,
};
