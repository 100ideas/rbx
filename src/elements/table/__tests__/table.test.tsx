import { Table } from "src/elements/table/table";

import {
  hasProperties,
  makeGenericHOCShallowWrapperInContextConsumer,
  makeNodeFactory,
  testForwardRefAsExoticComponentIntegration,
  testThemeIntegration,
  validateBoolPropType,
} from "src/__tests__/testing";

const COMPONENT = Table;
const COMPONENT_NAME = "Table";
const DEFAULT_ELEMENT = "table";
const BULMA_CLASS_NAME = "table";

const makeNode = makeNodeFactory(COMPONENT);

describe(`${COMPONENT_NAME} component`, () => {
  hasProperties(COMPONENT, {
    defaultProps: { as: DEFAULT_ELEMENT },
  });

  testForwardRefAsExoticComponentIntegration(
    makeNode,
    makeGenericHOCShallowWrapperInContextConsumer,
    DEFAULT_ELEMENT,
    BULMA_CLASS_NAME,
  );

  testThemeIntegration(makeNode, makeGenericHOCShallowWrapperInContextConsumer);

  describe("props", () => {
    const { propTypes } = COMPONENT;

    describe("bordered", () => {
      validateBoolPropType(propTypes, "bordered");

      [false, true].map(bordered => {
        it(`should ${bordered ? "" : "not "}be bordered`, () => {
          const node = makeNode({ bordered });
          const wrapper = makeGenericHOCShallowWrapperInContextConsumer(node);
          expect(wrapper.hasClass("is-bordered")).toBe(bordered);
        });
      });
    });

    describe("fullwidth", () => {
      validateBoolPropType(propTypes, "fullwidth");

      [false, true].map(fullwidth => {
        it(`should ${fullwidth ? "" : "not "}be fullwidth`, () => {
          const node = makeNode({ fullwidth });
          const wrapper = makeGenericHOCShallowWrapperInContextConsumer(node);
          expect(wrapper.hasClass("is-fullwidth")).toBe(fullwidth);
        });
      });
    });

    describe("hoverable", () => {
      validateBoolPropType(propTypes, "hoverable");

      [false, true].map(hoverable => {
        it(`should ${hoverable ? "" : "not "}be hoverable`, () => {
          const node = makeNode({ hoverable });
          const wrapper = makeGenericHOCShallowWrapperInContextConsumer(node);
          expect(wrapper.hasClass("is-hoverable")).toBe(hoverable);
        });
      });
    });

    describe("narrow", () => {
      validateBoolPropType(propTypes, "narrow");

      [false, true].map(narrow => {
        it(`should ${narrow ? "" : "not "}be narrow`, () => {
          const node = makeNode({ narrow });
          const wrapper = makeGenericHOCShallowWrapperInContextConsumer(node);
          expect(wrapper.hasClass("is-narrow")).toBe(narrow);
        });
      });
    });

    describe("striped", () => {
      validateBoolPropType(propTypes, "striped");

      [false, true].map(striped => {
        it(`should ${striped ? "" : "not "}be striped`, () => {
          const node = makeNode({ striped });
          const wrapper = makeGenericHOCShallowWrapperInContextConsumer(node);
          expect(wrapper.hasClass("is-striped")).toBe(striped);
        });
      });
    });
  });
});
