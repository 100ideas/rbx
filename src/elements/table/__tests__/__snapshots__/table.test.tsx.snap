// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component should concat classname in props with Bulma classname 1`] = `
<table
  className="table other-class this-is-a-test"
>
  <p>
    Default
  </p>
</table>
`;

exports[`Table component should exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "bordered": false,
    "fullwidth": false,
    "hoverable": false,
    "narrow": false,
    "striped": false,
  },
  "render": [Function],
}
`;

exports[`Table component should have table classname 1`] = `
<table
  className="table"
>
  Test 
  <a
    href="test"
  >
    Give me
  </a>
</table>
`;

exports[`Table component should use inline styles 1`] = `
<table
  className="table"
  style={
    Object {
      "height": 250,
    }
  }
>
  <p>
    Default
  </p>
</table>
`;
